File: Python/programmers/kakao/level5/시험장_나누기.py
import sys
sys.setrecursionlimit(10**6)

def exam_room(node, key, num, links):
    global group
    if node == -1:
        return 0
    left_cnt = exam_room(links[node][0], key, num, links)
    right_cnt = exam_room(links[node][1], key, num, links)
    if left_cnt + right_cnt + num[node] <= key:
        return left_cnt + right_cnt + num[node]
    else:
        if left_cnt + num[node] > key and right_cnt + num[node] <= key:
            group += 1
            return right_cnt + num[node]
        if left_cnt + num[node] <= key and right_cnt + num[node] > key:
            group += 1
            return left_cnt + num[node]
        if left_cnt + num[node] > key and right_cnt + num[node] > key:
            group += 2
            return num[node]
        if left_cnt + num[node] <= key and right_cnt + num[node] <= key:
            group += 1
            return min(left_cnt, right_cnt) + num[node]

group = 1
def solution(k, num, links):
    global group
    parent = [-1]*len(num)
    for node, (a, b) in enumerate(links):
        if a != -1:
            parent[a] = node
        if b != -1:
            parent[b] = node
    root_node = parent.index(-1)
    
    if k > exam_room(root_node, max(num), num, links):
        return max(num)

    lo, hi = max(num), sum(num)
    while lo < hi:
        group = 1
        mid = (lo+hi)//2
        exam_room(root_node, mid, num, links)
        if group > k:
            lo = mid+1
        else:
            hi = mid
    return lo


k = 3
num = [12, 30, 1, 8, 8, 6, 20, 7, 5, 10, 4, 1]
links = [[-1, -1], [-1, -1], [-1, -1], [-1, -1], [8, 5], [2, 10], [3, 0], [6, 1], [11, -1], [7, 4], [-1, -1], [-1, -1]]
print(solution(k, num, links)) # 40

k = 1
num = [6, 9, 7, 5]
links = [[-1, -1], [-1, -1], [-1, 0], [2, 1]]
print(solution(k, num, links)) # 27

k = 2
num = [6, 9, 7, 5]
links = [[-1, -1], [-1, -1], [-1, 0], [2, 1]]
print(solution(k, num, links)) # 14

k = 3
num = [6, 9, 7, 5]
links = [[-1, -1], [-1, -1], [-1, 0], [2, 1]]
print(solution(k, num, links)) # 9

File: Java/programmers/kakao/level5/시험장나누기.java
package CodeTest.Java.programmers.kakao.level5;

import java.util.*;

public class 시험장나누기 {
    private static int group = 1;

    public int solution(int k, int[] num, int[][] links) {
        int[] parent = new int[num.length];
        Arrays.fill(parent, -1);
        
        for (int i=0; i<links.length; i++) {
            if (links[i][0] != -1) parent[links[i][0]] = i;
            if (links[i][1] != -1) parent[links[i][1]] = i;
        }

        int root = -1;
        for (int i=0; i<parent.length; i++) {
            if (parent[i] == -1) {
                root = i;
                break;
            }
        }

        int _max = 0, sum = 0;
        for (int i=0; i<num.length; i++) {
            _max = Math.max(_max, num[i]);
            sum += num[i];
        }

        if (k > exam_room(root, _max, num, links)) return _max;

        int answer = 0;
        int low = _max, high = sum;
        while (low <= high) {
            group = 1;
            int mid = (low+high)/2;
            exam_room(root, mid, num, links);
            if (group <= k) {
                high = mid-1;
                answer = mid;
            }
            else low = mid+1;
        }
        
        return answer;
    }

    private static int exam_room(int node, int key, int[] num, int[][] links) {
        if (node == -1) return 0;
        
        int left_cnt = exam_room(links[node][0], key, num, links);
        int right_cnt = exam_room(links[node][1], key, num, links);

        if (left_cnt+right_cnt+num[node] <= key) {
            return left_cnt+right_cnt+num[node];
        } else {
            if (left_cnt+num[node] > key && right_cnt+num[node] <= key) {
                group += 1;
                return right_cnt+num[node];
            }
            if (left_cnt+num[node] <= key && right_cnt+num[node] > key) {
                group += 1;
                return left_cnt+num[node];
            }
            if (left_cnt+num[node] > key && right_cnt+num[node] > key) {
                group += 2;
                return num[node];
            }
            if (left_cnt+num[node] <= key && right_cnt+num[node] <= key) {
                group += 1;
                return Math.min(left_cnt, right_cnt) + num[node];
            }
        }

        return -1;
    }
}
